apps/wallet/src/engine/runners/solana/history/index.ts

```diff
@@ -0,0 +1,428 @@
+	const status = !meta?.err ? 'Success' : 'Failed';
+	const tokenForFee = getTransactionTokenFee();
+	const fee = (meta?.fee || 0) / 10 ** 9;
+	const balances = await getTransactionBalances(
+		connection,
+		parsedTransaction,
+		walletAddress,
+	);
+
+	if (!tokenForFee || balances.transactionType === 'Unknown') {
+		return {
+			_id,
+			id,
+			type: 'History',
+			signature,
+			network: Networks.solana,
+			transactionType: 'Unknown',
+			status,
+			date,
+		};
+	}
+
+	return {
+		_id,
+		type: 'History',
+		id,
+		signature,
+		network: Networks.solana,
+		tokenForFee,
+		fee,
+		date,
+		status,
+		...balances,
+	};
+};
+
+const getTransferTxType = (
+	walletAddress: string,
+	sender?: string,
+): 'Sent' | 'Received' => {
+	return sender === walletAddress ? 'Sent' : 'Received';
+};
+
+const getTransactionBalances = async (
+	connection: Connection,
+	parsedTransaction: ParsedTransactionWithMeta,
+	ownerAddress: string,
+): Promise<TransactionBalances> => {
+	const { meta, transaction } = parsedTransaction;
+
+	if (!meta || !meta?.preTokenBalances || !meta?.postTokenBalances)
+		return { transactionType: 'Unknown' };
+
+	const isJupiterSwap = meta?.logMessages?.find((message) =>
+		message.includes(JUPITER_V6_PROGRAM_ID),
+	);
+
+	if (isJupiterSwap) {
+		return await parseSwapTransaction({ transaction, meta, ownerAddress });
+	}
+
+	const isSplTokenTransaction =
+		meta.preTokenBalances?.length > 0 || meta.postTokenBalances?.length > 0;
+
+	if (isSplTokenTransaction) {
+		return await getSplTransactionBalances(
+			connection,
+			meta.preTokenBalances,
+			meta.postTokenBalances,
+			ownerAddress,
+		);
+	} else {
+		return await getNativeTransactionBalances(transaction, ownerAddress);
+	}
+};
+
+const parseSwapTransaction = async ({
+	transaction,
+	meta,
+	ownerAddress,
+}: {
+	transaction: ParsedTransaction;
+	meta: ParsedTransactionMeta;
+	ownerAddress: string;
+}): Promise<SolanaSwapHistory | { transactionType: 'Unknown' }> => {
+	const {
+		receivedTokenMint,
+		receivedTokenAmount,
+		sentTokenMint,
+		sentTokenAmount,
+	} = extractSwapBalance(transaction, meta);
+
+	if (!receivedTokenMint || !sentTokenMint)
+		return { transactionType: 'Unknown' };
+
+	const isWrappedSol =
+		receivedTokenMint === 'So11111111111111111111111111111111111111112';
+
+	const receivedTokenId = `${ownerAddress}/token/${
+		isWrappedSol ? solMint : receivedTokenMint
+	}`;
+	const sentTokenId = `${ownerAddress}/token/${sentTokenMint}`;
+
+	const [receivedTokenDoc, sentTokenDoc] = await Promise.all([
+		storage.safeGet<TokenDocument<SolanaToken>>(receivedTokenId),
+		storage.safeGet<TokenDocument<SolanaToken>>(sentTokenId),
+	]);
+
+	const receivedToken = {
+		amount: receivedTokenAmount / 10 ** (receivedTokenDoc?.decimals || 9),
+		metadata: {
+			mint: receivedTokenMint,
+			name: receivedTokenDoc?.name || '',
+			symbol: receivedTokenDoc?.symbol || '',
+			image: receivedTokenDoc?.image || '',
```

Comment:
could use "Unknown" here for `name` and `symbol`, I applied this default for token and nft also